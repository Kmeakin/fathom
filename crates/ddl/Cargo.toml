[package]
name = "ddl"
version = "0.1.0"
authors = ["Brendan Zabarauskas <bjzaba@yahoo.com.au>"]
edition = "2018"
publish = false

description = "A language for declaratively specifying binary data formats."
readme = "../../README.md"
license = "Apache-2.0"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
codespan = { git = "https://github.com/brendanzab/codespan", rev = "0a5d9a0daa1baf3142b119df56c19a82b7c2a46a" }
codespan-reporting = { git = "https://github.com/brendanzab/codespan", rev = "0a5d9a0daa1baf3142b119df56c19a82b7c2a46a" }
ddl-rt = { version = "0.1.0", path="../ddl-rt" }
itertools = "0.8"
lalrpop-util = "0.17"
pretty = "0.5"

[build-dependencies]
lalrpop = "0.17"
