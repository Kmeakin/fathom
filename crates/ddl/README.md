# ddl

A language for declaratively specifying binary data formats.
