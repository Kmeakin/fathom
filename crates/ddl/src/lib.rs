#![warn(rust_2018_idioms)]

pub mod compile;
pub mod concrete;
pub mod core;
pub mod elaborate;
pub mod parse;
