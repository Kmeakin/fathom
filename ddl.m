:- module ddl.

% Copyright (C) 2014-2015 YesLogic Pty. Ltd.
% All rights reserved.

:- interface.

:- import_module list, assoc_list, maybe, string, char.

:- type ddl == assoc_list(string, ddl_def).

:- type ddl_def
    --->    ddl_def(
		def_name :: string,
		def_args :: list(string),
		def_type :: ddl_type
	    ).

:- type ddl_type
    --->    ddl_type_word(word_type, word_values)
    ;	    ddl_type_array(expr, ddl_type)
    ;	    ddl_type_zero_or_more(ddl_type)
    ;	    ddl_type_sized(ddl_type, expr)
    ;	    ddl_type_struct(list(field_def))
    ;	    ddl_type_union(list(string))
    ;	    ddl_type_named(string, list(string))
    ;	    ddl_type_tag_magic(string).

:- type field_def
    --->    field_def(
		field_name :: string,
		field_type :: ddl_type,
		field_cond :: maybe(expr)
	    ).

:- type expr
    --->    expr_field(string)
    ;	    expr_const(int)
    ;	    expr_op(expr_op, expr, expr).

:- type expr_op
    --->    expr_add
    ;	    expr_sub
    ;	    expr_mul
    ;	    expr_and.

:- type word_values
    --->    word_values(
		word_values_any :: maybe(word_interp),
		word_values_enum :: assoc_list(word_value, word_interp)
	    ).

:- type word_type
    --->    uint8
    ;	    uint16
    ;	    uint32
    ;	    int8
    ;	    int16
    ;	    int32.

:- type word_value
    --->    word_value_int(int)
    ;	    word_value_tag(char, char, char, char).

:- type word_interp
    --->    word_interp_none
    ;	    word_interp_error
    ;	    word_interp_offset(offset).

:- type offset
    --->    offset(
		offset_base :: offset_base,
		offset_type :: ddl_type
	    ).

:- type offset_base
    --->    offset_base_struct
    ;	    offset_base_root
    ;	    offset_base_named(string).

:- type scope == assoc_list(string, int).

:- func word_type_size(word_type) = int.

:- func ddl_def_size(ddl, scope, ddl_def, list(int)) = int.

:- func union_options_size(ddl, scope, list(string)) = int.

:- func struct_fields_size(ddl, scope, list(field_def)) = int.

:- func ddl_type_size(ddl, scope, ddl_type) = int.

:- func eval_expr(scope, expr) = int.

:- func tag_num_to_string(int) = string.

:- func scope_resolve(scope, string) = int.

:- pred ddl_search(ddl::in, string::in, ddl_def::out) is semidet.

:- func ddl_lookup_det(ddl, string) = ddl_def.

%--------------------------------------------------------------------%

:- implementation.

:- import_module int.

:- import_module abort.

word_type_size(uint8) = 1.
word_type_size(uint16) = 2.
word_type_size(uint32) = 4.
word_type_size(int8) = 1.
word_type_size(int16) = 2.
word_type_size(int32) = 4.

ddl_def_size(DDL, Scope, Def, _Args) = Size :-
    Size = ddl_type_size(DDL, Scope, Def ^ def_type).

union_options_size(_DDL, _Scope, _Opts) = _Size :-
    abort("uh oh!").

struct_fields_size(_DDL, _Scope, []) = 0.
struct_fields_size(DDL, Scope, [F|Fs]) = Size :-
    Size0 = ddl_type_size(DDL, Scope, F ^ field_type),
    Size1 = struct_fields_size(DDL, Scope, Fs),
    Size = Size0 + Size1.

ddl_type_size(DDL, Scope, Type) = Size :-
    (
	Type = ddl_type_word(Word, _Values),
	Size = word_type_size(Word)
    ;
	Type = ddl_type_array(SizeExpr, Type0),
	Length = eval_expr(Scope, SizeExpr),
	Size = Length * ddl_type_size(DDL, Scope, Type0)
    ;
	Type = ddl_type_zero_or_more(_Type0),
	abort("FIXME zero_or_more has undefined size")
    ;
	Type = ddl_type_sized(_Type0, SizeExpr),
	Size = eval_expr(Scope, SizeExpr)
    ;
	Type = ddl_type_struct(Fields),
	Size = struct_fields_size(DDL, Scope, Fields)
    ;
	Type = ddl_type_union(Options),
	Size = union_options_size(DDL, Scope, Options)
    ;
	Type = ddl_type_named(Name, Args0),
	Def = ddl_lookup_det(DDL, Name),
	Args = map(scope_resolve(Scope), Args0),
	Size = ddl_def_size(DDL, Scope, Def, Args)
    ;
	Type = ddl_type_tag_magic(Name),
	TagNum = scope_resolve(Scope, Name),
	TagStr = tag_num_to_string(TagNum),
	Def = ddl_lookup_det(DDL, TagStr),
	Size = ddl_def_size(DDL, Scope, Def, [])
    ).

eval_expr(Scope, Expr) = Res :-
    (
	Expr = expr_field(Name),
	Res = scope_resolve(Scope, Name)
    ;
	Expr = expr_const(Res)
    ;
	Expr = expr_op(Op, Lhs0, Rhs0),
	Lhs = eval_expr(Scope, Lhs0),
	Rhs = eval_expr(Scope, Rhs0),
	(
	    Op = expr_add,
	    Res = Lhs + Rhs
	;
	    Op = expr_sub,
	    Res = Lhs - Rhs
	;
	    Op = expr_mul,
	    Res = Lhs * Rhs
	;
	    Op = expr_and,
	    Res = Lhs /\ Rhs
	)
    ).

tag_num_to_string(N) = Tag :-
    B0 = (N >> 24) /\ 0xFF,
    B1 = (N >> 16) /\ 0xFF,
    B2 = (N >> 8) /\ 0xFF,
    B3 = N /\ 0xFF,
    ( if
	char.from_int(B0, C0),
	char.from_int(B1, C1),
	char.from_int(B2, C2),
	char.from_int(B3, C3)
    then
	Tag = string.from_char_list([C0,C1,C2,C3])
    else
	abort("not a tag: "++int_to_string(N))
    ).

scope_resolve(Scope, Name) = Value :-
    ( if search(Scope, Name, Value0) then
	Value = Value0
    else
	abort("scope missing: "++Name)
    ).

ddl_search(DDL, Name, Def) :-
    search(DDL, Name, Def).

ddl_lookup_det(DDL, Name) = Def :-
    ( if search(DDL, Name, Def0) then
	Def = Def0
    else
	abort("unknown type: "++Name)
    ).

%--------------------------------------------------------------------%

:- end_module ddl.

