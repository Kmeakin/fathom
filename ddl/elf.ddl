
// FIXME endianness differs

// @root
ELF: struct {
    magic: byte[4], // FIXME = [0x7F, 'E', 'L', 'F'],
    class: byte = 1 | 2, // 32-bit or 64-bit
    data: byte = 1 | 2, // little or big endian
    version1: byte, // FIXME = 1 ?
    os_abi: byte, // FIXME
    abi_version: byte, // FIXME unused on Linux
    pad: byte[7],
    type: uint16 = 1 | 2 | 3 | 4, // relocatable, executable, shared, core
    machine: uint16, // target instruction set architecture
    version2: byte // FIXME version again???
}
