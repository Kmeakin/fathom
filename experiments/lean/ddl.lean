namespace ddl

  -- SYNTAX

  inductive endianness : Type
    | little
    | big

  mutual inductive expr, type, kind
    with expr : Type
      | true : expr
      | false : expr
      | nat : ℕ → expr
      | add : expr → expr → expr
      | mul : expr → expr → expr
      | var : string → expr
      | interp : expr → expr

    with type : Type
      | bool : type
      | nat : type
      | u8 : type
      | u16 : endianness → type
      | var : string → type
      | abs : string → kind → type → type
      | app : type → type → type
      | struct : string → type → type → type
      | array : type → expr → type

    with kind : Type
      | host : kind
      | binary : kind
      | arrow : kind → kind → kind

  infix + := expr.add
  infix * := expr.mul
  notation `Λ` x `:` κ `,` τ := type.abs x κ τ
  notation κ₁ ` ↣ ` κ₂ := kind.arrow κ₁ κ₂


  -- ENVIRONMENTS

  inductive env : Type
    | empty : env
    | cons : string × kind → env → env

  instance : has_emptyc env :=
    ⟨env.empty⟩

  instance : has_insert (string × kind) env :=
    ⟨env.cons⟩

  def env.mem : string × kind → env → Prop
    | (_, _) env.empty :=
        false
    | (x₁, b₁) (env.cons (x₂, b₂) Γ) :=
      if x₁ = x₂ then b₁ = b₂ else env.mem (x₁, b₁) Γ

  instance : has_mem (string × kind) env :=
    ⟨env.mem⟩


  -- RULES

  inductive type_of : env → expr → type → Type
    notation `τ[ ` Γ ` ⊢ ` e ` : ` τ ` ]` := type_of Γ e τ

    | true : Π {Γ},
        τ[ Γ ⊢ expr.true : type.bool ]

    | false : Π {Γ},
        τ[ Γ ⊢ expr.false : type.bool ]

    | nat : Π {Γ n},
        τ[ Γ ⊢ expr.nat n : type.nat ]

    | add : Π {Γ e₁ e₂},
        τ[ Γ ⊢ e₁ : type.nat ] →
        τ[ Γ ⊢ e₂ : type.nat ] →
        τ[ Γ ⊢ e₁ + e₂ : type.nat ]

    | mul : Π {Γ e₁ e₂},
        τ[ Γ ⊢ e₁ : type.nat ] →
        τ[ Γ ⊢ e₂ : type.nat ] →
        τ[ Γ ⊢ e₁ * e₂ : type.nat ]

    | interp_nat_to_u8 : Π {Γ e},
        τ[ Γ ⊢ e : type.nat ] →
        τ[ Γ ⊢ expr.interp e : type.u8 ]

    | interp_nat_to_u16 : Π {Γ e E},
        τ[ Γ ⊢ e : type.nat ] →
        τ[ Γ ⊢ expr.interp e : type.u16 E ]

    | interp_u8_to_nat : Π {Γ e},
        τ[ Γ ⊢ e : type.u8 ] →
        τ[ Γ ⊢ expr.interp e : type.nat ]

    | interp_u16_to_nat : Π {Γ e E},
        τ[ Γ ⊢ e : type.u16 E ] →
        τ[ Γ ⊢ expr.interp e : type.nat ]

  notation `τ[ ` Γ ` ⊢ ` e ` : ` τ ` ]` := type_of Γ e τ

  inductive kind_of : env → type → kind → Type
    notation `κ[ ` Γ ` ⊢ ` τ ` : ` κ ` ]` := kind_of Γ τ κ

    | bool : Π {Γ},
        κ[ Γ ⊢ type.bool : kind.host ]

    | nat : Π {Γ},
        κ[ Γ ⊢ type.nat : kind.host ]

    | u8 : Π {Γ},
        κ[ Γ ⊢ type.u8 : kind.binary ]

    | u16 : Π {Γ E},
        κ[ Γ ⊢ type.u16 E : kind.binary ]

    | var : Π {Γ x κ},
        (x, κ) ∈ Γ →
        κ[ Γ ⊢ type.var x : κ ]

    | abs : Π {Γ x τ₁ κ₁ κ₂},
        κ[ (insert (x, κ₁) Γ) ⊢ τ₁ : κ₂ ] →
        κ[ Γ ⊢ Λ x : κ₁, τ₁ : (κ₁ ↣ κ₂) ]

    | app : Π {Γ τ₁ τ₂ κ₁ κ₂},
        κ[ Γ ⊢ τ₁ : (κ₁ ↣ κ₂) ] →
        κ[ Γ ⊢ τ₂ : κ₁ ] →
        κ[ Γ ⊢ type.app τ₁ τ₂ : κ₂ ]

    -- arrays take on the kind of their elements
    -- size expressions must always evaluate to natural numbers
    | array : Π {Γ τ κ e},
        κ[ Γ ⊢ τ : κ ] →
        τ[ Γ ⊢ e : type.nat ] →
        κ[ Γ ⊢ type.array τ e : κ ]

    -- structs are always binary, and subsequent fields can
    -- access previous fields
    | struct : Π {Γ x τ₁ τ₂},
        κ[ Γ ⊢ τ₁ : kind.binary ] →
        κ[ insert (x, kind.binary) Γ ⊢ τ₂ : kind.binary ] →
        κ[ Γ ⊢ type.struct x τ₁ τ₂ : kind.binary ]

  notation `κ[ ` Γ ` ⊢ ` τ ` : ` κ ` ]` := kind_of Γ τ κ


  section
    open endianness expr type kind

    example : κ[ {("x", binary)} ⊢  array (var "x") (expr.nat 1) : binary ] :=
      kind_of.array (kind_of.var rfl) type_of.nat

    example : κ[ ∅ ⊢  Λ "x" : binary, array (var "x") (expr.nat 1 + expr.nat 2) : (binary ↣ binary) ] :=
      kind_of.abs (kind_of.array (kind_of.var rfl)
                  (type_of.add type_of.nat type_of.nat))
  end

  -- PROOFS

  -- TODO

end ddl
