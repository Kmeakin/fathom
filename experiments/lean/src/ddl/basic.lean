namespace ddl

  def relation (α : Type) :=
    α → α → Prop

  inductive multi {α : Type} (ρ : relation α) : relation α
    | refl {x : α} : multi x x
    | step {x y z : α} : ρ x y → multi y z → multi x z

end ddl
