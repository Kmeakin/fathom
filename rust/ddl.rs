#![allow(dead_code)]
#![allow(unused_variables)]

enum ExprBool {
    Or(Box<ExprBool>, Box<ExprBool>),
    Eq(Box<ExprInt>, Box<ExprInt>),
    Gt(Box<ExprInt>, Box<ExprInt>)
}

enum ExprInt {
    Const(u32),
    Field(String)
}

enum Type {
    Int(IntType),
    Struct(Box<StructType>),
    Array(Box<ArrayType>)
}

enum IntType {
    Uint8,
    Uint16,
    Uint32,
    Int8,
    Int16,
    Int32
}

struct StructType {
    items: Vec<StructItem>,
    conds: Vec<ExprBool>
}

enum StructItem {
    Field(Field),
    CondSection(ExprBool, StructType)
}

struct Field {
    field_name: String,
    field_type: Type
}

struct ArrayType {
    array_length: ExprInt,
    array_type: Type
}

fn match_type(buf: &[u8], ddl_type: Type) -> bool {
    false // FIXME not yet implemented
}

fn main() {
    let f1 = Field {
        field_name: String::from("version"),
        field_type: Type::Int(IntType::Uint8)
    };

    let f2 = Field {
        field_name: String::from("length"),
        field_type: Type::Int(IntType::Uint8)
    };

    let atype = ArrayType {
        array_length: ExprInt::Field(String::from("length")),
        array_type: Type::Int(IntType::Uint8)
    };

    let f3 = Field {
        field_name: String::from("values"),
        field_type: Type::Array(Box::new(atype))
    };

    let f4 = Field {
        field_name: String::from("extra"),
        field_type: Type::Int(IntType::Uint8)
    };

    let c1 =
        ExprBool::Or(
            Box::new(ExprBool::Eq(
                Box::new(ExprInt::Field(String::from("version"))),
                Box::new(ExprInt::Const(1))
            )),
            Box::new(ExprBool::Eq(
                Box::new(ExprInt::Field(String::from("version"))),
                Box::new(ExprInt::Const(2))
            ))
        );

    let c2 =
        ExprBool::Gt(
            Box::new(ExprInt::Field(String::from("version"))),
            Box::new(ExprInt::Const(1))
        );

    let s1 = StructType {
        items: vec![StructItem::Field(f4)],
        conds: vec![]
    };

    let s = StructType {
        items: vec![
            StructItem::Field(f1),
            StructItem::Field(f2),
            StructItem::Field(f3),
            StructItem::CondSection(c2, s1)
        ],
        conds: vec![c1]
    };

    let ddl = Type::Struct(Box::new(s));

    let buf: [u8; 0] = []; // FIXME load from file

    let b = match_type(&buf, ddl);

    println!("match result: {}", b);
}

