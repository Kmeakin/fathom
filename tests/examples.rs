extern crate ddl;

use ddl::Env;

#[test]
#[ignore]
fn cmap() {
    const SRC: &str = include_str!("../examples/ddl/cmap.ddl");

    let mut env = Env::default();
    let defs = ddl::parser::parse(&env, SRC).unwrap();
    env.check_defs(defs).unwrap();
}

#[test]
#[ignore]
fn edid() {
    const SRC: &str = include_str!("../examples/ddl/edid.ddl");

    let mut env = Env::default();
    let defs = ddl::parser::parse(&env, SRC).unwrap();
    env.check_defs(defs).unwrap();
}

#[test]
fn heroes_of_might_and_magic_bmp() {
    const SRC: &str = include_str!("../examples/ddl/heroes_of_might_and_magic_bmp.ddl");

    let mut env = Env::default();
    let defs = ddl::parser::parse(&env, SRC).unwrap();
    env.check_defs(defs).unwrap();
}

#[test]
#[ignore]
fn object_id() {
    const SRC: &str = include_str!("../examples/ddl/object_id.ddl");

    let mut env = Env::default();
    let defs = ddl::parser::parse(&env, SRC).unwrap();
    env.check_defs(defs).unwrap();
}
