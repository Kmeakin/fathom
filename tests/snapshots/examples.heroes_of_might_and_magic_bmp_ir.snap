Program {
    defs: {
        Path {
            base: "HerosOfMightAndMagicBmp",
            children: []
        }: Struct(
            [
                Field {
                    name: "magic",
                    value: Const(
                        Unsigned(
                            U16
                        )
                    )
                },
                Field {
                    name: "width",
                    value: Const(
                        Unsigned(
                            U16
                        )
                    )
                },
                Field {
                    name: "height",
                    value: Const(
                        Unsigned(
                            U16
                        )
                    )
                },
                Field {
                    name: "data",
                    value: Array(
                        Const(
                            Unsigned(
                                U8
                            )
                        )
                    )
                }
            ],
            Some(
                Sequence(
                    [
                        Named("magic", Const(
                            U16Le
                        )),
                        Named("width", Const(
                            U16Le
                        )),
                        Named("height", Const(
                            U16Le
                        )),
                        Named("data", Repeat(
                            Const(
                                U8
                            ),
                            Exact(
                                Binop(
                                    Mul,
                                    Var(
                                        Bound(Named("width", BoundVar(1, 0)))
                                    ),
                                    Var(
                                        Bound(Named("height", BoundVar(0, 0)))
                                    )
                                )
                            )
                        ))
                    ],
                    Struct(
                        Path {
                            base: "HerosOfMightAndMagicBmp",
                            children: []
                        },
                        [
                            Field {
                                name: "magic",
                                value: Var(
                                    Bound(Named("magic", BoundVar(3, 0)))
                                )
                            },
                            Field {
                                name: "width",
                                value: Var(
                                    Bound(Named("width", BoundVar(2, 0)))
                                )
                            },
                            Field {
                                name: "height",
                                value: Var(
                                    Bound(Named("height", BoundVar(1, 0)))
                                )
                            },
                            Field {
                                name: "data",
                                value: Var(
                                    Bound(Named("data", BoundVar(0, 0)))
                                )
                            }
                        ]
                    )
                )
            )
        )
    }
}
